### ISSUE TITLE
where are text_ids、text_labels、text_masks of infer() of VLMo class from?

### ISSUE THREAD

**Issue by yangzhj53** [microsoft/unilm#1360]
I wanna use VLMo to inference on NLVR task, but I'm confused about these:
        text_ids = batch[f"text_ids{do_mlm}"]
        text_labels = batch[f"text_labels{do_mlm}"]
        text_masks = batch[f"text_masks"]
I read the code carefully, but I have not found where they come from.

**Comment by yangzhj53** [comment]
In my opinion, datasets are created by the function:

    def dataset_cls(self):
        raise NotImplementedError("return tuple of dataset class")

But it have not been implemented yet, so the input dataset is not aligned with the dataset used for the inference.

**Comment by biomolecule** [comment]
> In my opinion, datasets are created by the function:
> 
> ```
> def dataset_cls(self):
>     raise NotImplementedError("return tuple of dataset class")
> ```
> 
> But it have not been implemented yet, so the input dataset is not aligned with the dataset used for the inference.

### **They are generated by function called `collate` which is located in `/vlmo/datasets/BaseDataset.py`**

The following shows how to find it.

The datamodule of VLMo is defined in `/vlmo/datamodules/multitask_datamodule.py`, which "concat" a series of "unitask_datamodule" in the `self.dms`. In your case, the "unitask_datamodule" comes from `/vlmo/datamodules/nlvr2_datamodule.py`, a child class of `BaseDataModule` in  `/vlmo/datamodules/BaseDataModule.py`. 

Obviously, the datamodule loads and preprocesses the data from the dataset. In `/vlmo/datamodules/BaseDataModule.py`, method `set_train_dataset` set which dataset should be loaded through `self.dataset_cls()`. Actually it is defined in the child class `NLVR2DataModule`, called `NLVR2Dataset`, which is imported from `vlmo.datasets`. Also you kan get the original data.

         return {
            "image_0": image_tensor_0,
            "image_1": image_tensor_1,
            "text": text,
            "answers": answers,
            "table_name": self.table_names[index],
        }

The ids, labels, masks and other information is created by the param `collate_fn` in dataloader. Go back into `/vlmo/datamodules/BaseDataModule.py`, in `train_dataloader`, you can find such param, which process the data in batch before the dataloader feed them to the model.

The `/vlmo/datamodules/BaseDataModule/train_dataloader`, the `collate_fn` is setted as `self.train_dataset.collate`, and the `self.train_dataset` is an istance of `NLVR2Dataset`. And `NLVR2Dataset` is also the child class of `BaseDataset`. Don't forget our goal is to find the method "collate", wthichis loacted in the end of `BaseDataset` in `/vlmo/datasets/BaseDataset.py`.

Such function read-in bath_size of data, in a list, and do the other things, such as tokenize, wordmasking, and transform each kind of data in list into a large tensor, which will be feed into the model.

    def collate(self, batch, mlm_collator):
        batch_size = len(batch)
        keys = set([key for b in batch for key in b.keys()])
        dict_batch = {k: [dic[k] if k in dic else None for dic in batch] for k in keys}
        img_keys = [k for k in list(dict_batch.keys()) if "image" in k]
        for img_key in img_keys:
            new_imgs = [tmp_img[0] for tmp_img in dict_batch[img_key]]
            batch_new_imgs = torch.stack(new_imgs, dim=0)
            dict_batch[img_key] = [batch_new_imgs]

        txt_keys = [k for k in list(dict_batch.keys()) if "text" in k]
        if len(txt_keys) != 0:
            texts = [[d[0] for d in dict_batch[txt_key]] for txt_key in txt_keys]
            encodings = [[d[1] for d in dict_batch[txt_key]] for txt_key in txt_keys]
            draw_text_len = len(encodings)
            flatten_encodings = [e for encoding in encodings for e in encoding]
            flatten_mlms = mlm_collator(flatten_encodings)

            for i, txt_key in enumerate(txt_keys):
                texts, encodings = (
                    [d[0] for d in dict_batch[txt_key]],
                    [d[1] for d in dict_batch[txt_key]],
                )

                mlm_ids, mlm_labels = (
                    flatten_mlms["input_ids"][batch_size * (i) : batch_size * (i + 1)],
                    flatten_mlms["labels"][batch_size * (i) : batch_size * (i + 1)],
                )

                input_ids = torch.zeros_like(mlm_ids)
                attention_mask = torch.zeros_like(mlm_ids)
                for _i, encoding in enumerate(encodings):
                    _input_ids, _attention_mask = (
                        torch.tensor(encoding["input_ids"]),
                        torch.tensor(encoding["attention_mask"]),
                    )
                    input_ids[_i, : len(_input_ids)] = _input_ids
                    attention_mask[_i, : len(_attention_mask)] = _attention_mask

                dict_batch[txt_key] = texts
                dict_batch[f"{txt_key}_ids"] = input_ids
                dict_batch[f"{txt_key}_labels"] = torch.full_like(input_ids, -100)
                dict_batch[f"{txt_key}_ids_mlm"] = mlm_ids
                dict_batch[f"{txt_key}_labels_mlm"] = mlm_labels
                dict_batch[f"{txt_key}_masks"] = attention_mask

        return dict_batch

**Comment by yangzhj53** [comment]
> > In my opinion, datasets are created by the function:
> > ```
> > def dataset_cls(self):
> >     raise NotImplementedError("return tuple of dataset class")
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > But it have not been implemented yet, so the input dataset is not aligned with the dataset used for the inference.
> 
> ### **They are generated by function called `collate` which is located in `/vlmo/datasets/BaseDataset.py`**
> The following shows how to find it.
> 
> The datamodule of VLMo is defined in `/vlmo/datamodules/multitask_datamodule.py`, which "concat" a series of "unitask_datamodule" in the `self.dms`. In your case, the "unitask_datamodule" comes from `/vlmo/datamodules/nlvr2_datamodule.py`, a child class of `BaseDataModule` in `/vlmo/datamodules/BaseDataModule.py`.
> 
> Obviously, the datamodule loads and preprocesses the data from the dataset. In `/vlmo/datamodules/BaseDataModule.py`, method `set_train_dataset` set which dataset should be loaded through `self.dataset_cls()`. Actually it is defined in the child class `NLVR2DataModule`, called `NLVR2Dataset`, which is imported from `vlmo.datasets`. Also you kan get the original data.
> 
> ```
>      return {
>         "image_0": image_tensor_0,
>         "image_1": image_tensor_1,
>         "text": text,
>         "answers": answers,
>         "table_name": self.table_names[index],
>     }
> ```
> 
> The ids, labels, masks and other information is created by the param `collate_fn` in dataloader. Go back into `/vlmo/datamodules/BaseDataModule.py`, in `train_dataloader`, you can find such param, which process the data in batch before the dataloader feed them to the model.
> 
> The `/vlmo/datamodules/BaseDataModule/train_dataloader`, the `collate_fn` is setted as `self.train_dataset.collate`, and the `self.train_dataset` is an istance of `NLVR2Dataset`. And `NLVR2Dataset` is also the child class of `BaseDataset`. Don't forget our goal is to find the method "collate", wthichis loacted in the end of `BaseDataset` in `/vlmo/datasets/BaseDataset.py`.
> 
> Such function read-in bath_size of data, in a list, and do the other things, such as tokenize, wordmasking, and transform each kind of data in list into a large tensor, which will be feed into the model.
> 
> ```
> def collate(self, batch, mlm_collator):
>     batch_size = len(batch)
>     keys = set([key for b in batch for key in b.keys()])
>     dict_batch = {k: [dic[k] if k in dic else None for dic in batch] for k in keys}
>     img_keys = [k for k in list(dict_batch.keys()) if "image" in k]
>     for img_key in img_keys:
>         new_imgs = [tmp_img[0] for tmp_img in dict_batch[img_key]]
>         batch_new_imgs = torch.stack(new_imgs, dim=0)
>         dict_batch[img_key] = [batch_new_imgs]
> 
>     txt_keys = [k for k in list(dict_batch.keys()) if "text" in k]
>     if len(txt_keys) != 0:
>         texts = [[d[0] for d in dict_batch[txt_key]] for txt_key in txt_keys]
>         encodings = [[d[1] for d in dict_batch[txt_key]] for txt_key in txt_keys]
>         draw_text_len = len(encodings)
>         flatten_encodings = [e for encoding in encodings for e in encoding]
>         flatten_mlms = mlm_collator(flatten_encodings)
> 
>         for i, txt_key in enumerate(txt_keys):
>             texts, encodings = (
>                 [d[0] for d in dict_batch[txt_key]],
>                 [d[1] for d in dict_batch[txt_key]],
>             )
> 
>             mlm_ids, mlm_labels = (
>                 flatten_mlms["input_ids"][batch_size * (i) : batch_size * (i + 1)],
>                 flatten_mlms["labels"][batch_size * (i) : batch_size * (i + 1)],
>             )
> 
>             input_ids = torch.zeros_like(mlm_ids)
>             attention_mask = torch.zeros_like(mlm_ids)
>             for _i, encoding in enumerate(encodings):
>                 _input_ids, _attention_mask = (
>                     torch.tensor(encoding["input_ids"]),
>                     torch.tensor(encoding["attention_mask"]),
>                 )
>                 input_ids[_i, : len(_input_ids)] = _input_ids
>                 attention_mask[_i, : len(_attention_mask)] = _attention_mask
> 
>             dict_batch[txt_key] = texts
>             dict_batch[f"{txt_key}_ids"] = input_ids
>             dict_batch[f"{txt_key}_labels"] = torch.full_like(input_ids, -100)
>             dict_batch[f"{txt_key}_ids_mlm"] = mlm_ids
>             dict_batch[f"{txt_key}_labels_mlm"] = mlm_labels
>             dict_batch[f"{txt_key}_masks"] = attention_mask
> 
>     return dict_batch
> ```

I see. Thanks for your detailed response.
